backend/store/memory/event_store.go

```diff
@@ -0,0 +1,348 @@
+
+	// GetSilences returns all silences in the namespace. A nil slice with no error is
+	// returned if none were found.
+	GetSilences(ctx context.Context, namespace string) ([]*corev2.Silenced, error)
+
+	// GetSilencedsByCheck returns all silences for the given check . A nil
+	// slice with no error is returned if none were found.
+	GetSilencesByCheck(ctx context.Context, namespace, check string) ([]*corev2.Silenced, error)
+
+	// GetSilencedEntriesBySubscription returns all entries for the given
+	// subscription. A nil slice with no error is returned if none were found.
+	GetSilencesBySubscription(ctx context.Context, namespace string, subscriptions []string) ([]*corev2.Silenced, error)
+
+	// GetSilenceByName returns an entry using the given namespace and name. An
+	// error is returned if the entry is not found.
+	GetSilenceByName(ctx context.Context, namespace, name string) (*corev2.Silenced, error)
+
+	// UpdateSilence creates or updates a given silence.
+	UpdateSilence(ctx context.Context, silence *corev2.Silenced) error
+
+	// GetSilencesByName gets all the named silence entries.
+	GetSilencesByName(ctx context.Context, namespace string, names []string) ([]*corev2.Silenced, error)
+}
+
+func (m *memorydb) ReadEntry(namespace, entity, check string) *eventEntry {
+	key := strings.Join([]string{namespace, entity, check}, "\n")
+	emptyEntry := &eventEntry{}
+	result, _ := m.data.LoadOrStore(key, emptyEntry)
+	entry := result.(*eventEntry)
+	return entry
+}
+
+func (m *memorydb) WriteTo(ctx context.Context, s store.EventStore) error {
+	var storeErr error
+	ctx = store.NoMergeEventContext(ctx)
+	m.data.Range(func(key, value any) bool {
+		entry := value.(*eventEntry)
+		entry.Mu.Lock()
+		defer entry.Mu.Unlock()
+		if !entry.Dirty {
+			return true
+		}
+		eventBytes, err := snappy.Decode(nil, entry.EventBytes)
+		if err != nil {
+			// developer error, fatal
+			panic(err)
+		}
+		var event corev2.Event
+		if err := proto.Unmarshal(eventBytes, &event); err != nil {
+			// developer error, fatal
+			panic(err)
+		}
+		_, _, storeErr = s.UpdateEvent(ctx, &event)
+		if storeErr != nil {
+			return false
+		}
+		entry.Dirty = false
+		return true
+	})
+	return storeErr
+}
+
+type EventStore struct {
+	backingStore      store.EventStore
+	flushInterval     time.Duration
+	eventWriteLimiter *rate.Limiter
+	db                *memorydb
+	silenceStore      SilenceStoreI
+	flushC            chan struct{}
+}
+
+type EventStoreConfig struct {
+	BackingStore    store.EventStore
+	FlushInterval   time.Duration
+	EventWriteLimit rate.Limit
+	SilenceStore    SilenceStoreI
+}
+
+func NewEventStore(config EventStoreConfig) *EventStore {
+	return &EventStore{
+		backingStore:      config.BackingStore,
+		flushInterval:     config.FlushInterval,
+		db:                newMemoryDB(),
+		silenceStore:      config.SilenceStore,
+		flushC:            make(chan struct{}),
+		eventWriteLimiter: rate.NewLimiter(config.EventWriteLimit, int(config.EventWriteLimit/10)),
```

Comment:
Was curious to see golang.org/x/time/rate in action. Looks like it didn't get fully hooked up.